src/plugins.c

```diff
@@ -230,3 +341,40 @@ ly_plugins_init(void)
     }
     return ret;
 }
+
+API LY_ERR
+ly_plugins_add(const char *pathname)
+{
+    LY_ERR ret = LY_SUCCESS;
+    uint32_t types_count, extensions_count;
+
+    LY_CHECK_ARG_RET(NULL, pathname, LY_EINVAL);
+
+    /* works only in case a context exists */
+    pthread_mutex_lock(&plugins_guard);
+    if (!context_refcount) {
+        /* no context */
+        pthread_mutex_unlock(&plugins_guard);
+        LOGERR(NULL, LY_EDENIED, "To add a plugin, at least one context must exists.");
+        return LY_EDENIED;
+    }
+
+    /* remember the current plugins lists for recovery */
+    types_count = plugins_types.count;
+    extensions_count = plugins_extensions.count;
+
+    ret = ly_plugins_load_module(pathname);
+    if (ret) {
+        /* revert changes in the lists */
```

Comment:
because the reaction differs if the call comes from `ly_plugins_add()` (only the plugins related to the `.so` being loaded are removed) or from plugins initiation (`ly_plugins_init()` via `ly_plugins_insert_dir()`) when all the plugins are removed.

Reply:
I see, so please make it clear from the comment. At least something like `revert all the changes in the lists`.

Reply:
ok, done

Reply:
um, is it?

Reply:
I have added comment into `ly_plugins_insert_dir()`, the comment in `ly_plugins_add()` is fine, isn't it? Using 'all the changes' seems more confusing for me, it's like 'also some other changes besides those done here'.

Reply:
The meaning was supposed to be 'also the changes not reverted by default'. I am just trying to fix this so that I do not have the same question I had now, why is some cleanup performed outside the function that did the changes? Comment in `ly_plugins_insert_dir()` does not address that, the information is needed on this line.

Reply:
because there is a difference in what you want to cleanup. But ok, since it is a subset, I can revert the plugins loaded in the call of `ly_plugins_load_module()` inside it and in case of the initialization it will then remove the rest of plugins.